README
NCGI
====

  NCGI is a collection of Perl modules that assist authors in writing
  Common Gateway Interface (CGI) scripts. It is kind of like the standard
  'CGI' module but much lighter in code and features, has a different
  inteface and produces better output.

INSTALLATION

  To install this module type the following:

    perl Makefile.PL
    make
    make test
    make install

DEPENDENCIES

  This module requires these other modules and libraries:

    XML::API         (CPAN)
    debug            (CPAN)

  If you want to actually do any debugging with NCGI then you also need
  the following module:

    Log::Delta       (CPAN)

COPYRIGHT AND LICENCE

  Copyright (C) 2005 Mark Lawrence <nomad@null.net>

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.



Changes
commit 3f431c365495024d79b3f31444071c7df47c09f1 (master)
Author: Mark Lawrence <nomad@null.net>
Date:   Thu Nov 20 20:30:20 2008 +0100

    CPAN - Uploaded as v0.12.

commit 1bb939a976878d9bb34a22c9ab0d2d26a500e38b (master~1)
Author: Mark Lawrence <nomad@null.net>
Date:   Thu Nov 20 20:26:03 2008 +0100

    Updated example/ncgi.cgi to include some UTF8 strings.

commit 95cb24aaf46866964f1565829a230955611c4408 (master~2)
Author: Mark Lawrence <nomad@null.net>
Date:   Thu Nov 20 20:16:03 2008 +0100

    Force NCGI::Query->param to return undef when appropriate.
    
    If the parameter was not submitted by the browser, the param() method
    was doing a straight "return". Using that value in something like a
    hash results in an uneven number of elements. Ie the following:
    
        (a => val, b => NCGI->q->param('notexist'))
    
    was exactly the same as this:
    
        (a => val, b)

commit 2466f9040f1393129b98c6ea7a8a8ccc1fa9fdf0 (master~3)
Author: Mark Lawrence <nomad@null.net>
Date:   Thu Oct 9 16:48:23 2008 +0200

    Makefile.PL: Requires I18N::LangTags::Detect
    
    CPAN
        Uploaded as v0.11.

commit 1ee9903478b7fe507c8fb16b347b42988b2b2a9d (master~4)
Author: Mark Lawrence <nomad@null.net>
Date:   Thu Oct 9 16:32:54 2008 +0200

    Removed ./debian from git control. dh-make-perl is your friend.

commit 696c8d29582e2fe92ec1cf2dbcd64232691628d1 (master~5)
Author: Mark Lawrence <nomad@null.net>
Date:   Thu Oct 9 16:21:28 2008 +0200

    NCGI::Response - Fixed two calls to XML::API->_add with undefined content.

commit 97921149fa01bf49a20455369801dfa422d9c131 (master~6)
Author: Mark Lawrence <nomad@null.net>
Date:   Fri Oct 3 16:41:15 2008 +0200

    NCGI Bugfix: stop hiding warnings that occur within eval blocks.

commit fa65764e2f410d18371f33455204936b0bcb24a0 (master~7)
Author: Mark Lawrence <nomad@null.net>
Date:   Sun Sep 21 15:33:15 2008 +0200

    NCGI
        Removed useless exit() from the die handler
    
    CPAN
        Uploaded as version 0.10.

commit e40e5ce69f46f8bbe9700ecc2334ddd3b0883902 (master~8)
Author: Mark Lawrence <nomad@null.net>
Date:   Sat Sep 20 18:20:09 2008 +0200

    Fixed up broken tests for NCGI::Cookie

commit 88f493faef86a5031915da21ac2cf231c60ebd62 (master~9)
Author: Mark Lawrence <nomad@null.net>
Date:   Sun Sep 14 15:09:46 2008 +0200

    NCGI::Cookie::expire() now takes unix epoch instead of minutes.

commit e6783c5e497f70227ee0afb6744d9d0d7e35b907 (master~10)
Author: Mark Lawrence <nomad@null.net>
Date:   Sun Sep 14 15:05:43 2008 +0200

    NCGI::Response::Header support for running under HTTP::Server::Simple

commit 46fbbe735db7e44999bfa94dcacdbe5f4c5abdd0 (master~11)
Author: Mark Lawrence <nomad@null.net>
Date:   Tue Jun 3 12:38:07 2008 +0200

    Makefile.PL, debian/control
        Require XML::API v0.22.
    
        Removed Debian build dependency on libmodule-build.
    
    NCGI
        Improved storage and rendering of warnings.
    
    NCGI::Cookie
        Properly create multiple cookies as an array.
        Improved tests.
    
    NCGI::Response
        Convert all XML::API::XHTML uses to generic
        XML::API->new(doctype...) calls.
    
        Implement the missing ->rss support now that XML::API::RSS works.
    
    NCGI::Response::Header
        Complain if '_print' is used and inform to use '_send'.
    
    CPAN
        Uploaded as v0.09.

commit f4be623cb3d5b34fdf39156656c73635c2e4bd02 (master~12)
Author: Mark Lawrence <nomad@null.net>
Date:   Tue Mar 25 12:21:25 2008 +0100

    NCGI::Cookie
        Added Max-Age parameter. Fixes:
            http://rt.cpan.org/Ticket/Display.html?id=34337
        Module now tested.
    
    Makefile.PL, debian/control
        Require XML::API 0.19.
    
    CPAN
        Uploaded as v0.08.

commit aef8a57ccf837226edebf4e886688ca370844a94 (master~13)
Author: Mark Lawrence <nomad@null.net>
Date:   Thu Mar 20 00:16:59 2008 +0100

    NCGI
        New fast response method frespond()
        Cleaned up the WARN and DIE handlers a little
    
    NCGI::Cookie
        Privatised some methods and documented others.
    
    NCGI::Response::Header, NCGI::Singleton
        Documentation update.
    
    Tests
        Fixed t/NCGI.t to be clean (but not at all complete).
    
    CPAN
        Uploaded at v0.07

commit 444cad36bd6ac966650dccde81613ad4cb2969c0 (master~14)
Author: Mark Lawrence <nomad@null.net>
Date:   Thu Aug 23 12:09:43 2007 +0200

    Remove some old debugging from lib/NCGI.pm

commit a8705cbc1edbd3059676605fb268cba878daf0d3 (master~15)
Author: Mark Lawrence <nomad@null.net>
Date:   Thu Aug 23 12:09:24 2007 +0200

    Removed inc/Module/Install/* from git repo.

commit 273d27e856f5e4e3c0c6be922453cdc87371f596 (master~16)
Author: Mark Lawrence <nomad@null.net>
Date:   Wed Jul 4 16:27:40 2007 +0200

    Instead of checking specifically for MSIE or Lynx, we now check if
    HTTP_ACCEPT contains application/xhtml+xml and if not then return
    text/html.

commit 7b8a80487de29eab6c44a111878987ca00a0f593 (master~17)
Author: Mark Lawrence <nomad@null.net>
Date:   Wed Jul 4 16:06:52 2007 +0200

    Cleanup/improvement of the debugging and warnings handling.
    Send broken Internet Explorer text/html instead of application/xhtml+xml

commit dbc9776857803fd55c440a76d24d87aadb16ea7c (master~18)
Author: Mark Lawrence <nomad@null.net>
Date:   Mon May 14 14:05:36 2007 +0200

    VERSION to 0.07

commit 29b24c9362efc7a70856aeceaa5f593bc4d95e01 (master~19)
Author: Mark Lawrence <nomad@null.net>
Date:   Mon May 14 13:56:35 2007 +0200

    README cleanup. NCGI::Cookie version to 0.06.
    Finally have everything in git and can start using it properly.

commit c09f157fab2d49097514612a6f6cfe044a71dc20 (master~20)
Author: Mark Lawrence <nomad@null.net>
Date:   Mon May 14 13:54:31 2007 +0200

    NCGI::Response::Header cleanup and documentation

commit a504ad3bbb5c9c1b04de7a75cfbb246d87fdb8a1 (master~21)
Author: Mark Lawrence <nomad@null.net>
Date:   Mon May 14 13:51:56 2007 +0200


lib/NCGI/Cookie.pm
package NCGI::Cookie;
use strict;
use warnings;
use Carp qw(croak);
use Digest::MD5 qw(md5_hex);
use CGI::Util qw(escape unescape);
use overload '""' => \&_as_string, 'fallback' => 1;

our $VERSION = '0.12';

# ----------------------------------------------------------------------
# Class Functions
# ----------------------------------------------------------------------

#
# Return a HASHREF of Cookies recieved
#
sub fetch {
    my $src = shift || $ENV{HTTP_COOKIE} || '';
    my @pairs = split(/\;\s*/, $src);
    my $hashref = {};
    foreach my $sets (@pairs) {
        my ($key,$val) = split(/=/, $sets);
        if (exists($hashref->{unescape($key)})) {
            if (ref($hashref->{unescape($key)}) eq 'ARRAY') {
                push(@{$hashref->{unescape($key)}}, unescape($val));
            }
            else {
                $hashref->{unescape($key)} =
                    [$hashref->{unescape($key)}, unescape($val)];
            }
        }
        else {
            $hashref->{unescape($key)} = unescape($val);
        }
    }
    return $hashref;
}

# ----------------------------------------------------------------------
# Object Methods
# ----------------------------------------------------------------------

#
# This routine takes (name,value,epoch,path,domain) as arguments
# to set a cookie.
#
sub new {
    my $proto = shift;
    my $class = ref($proto) || $proto;
    my $self  = {
        name    => undef,
        value   => undef,
        expires => time + 60*60*24,
        max_age => undef,
        path    => '/',
        domain  => undef,
        @_,
    };
    bless ($self, $class);

    $self->{name}  = escape(_cookie_scrub($self->{name}));
    $self->{value} = $self->{value} ? escape(_cookie_scrub($self->{value})) :
                     md5_hex(md5_hex(time . {} . rand() . $$));

    return $self;
}

sub name {
    my $self = shift;
    $self->{name} = _cookie_scrub(shift) if (@_);
    return $self->{name};
}

sub value {
    my $self = shift;
    $self->{value} = _cookie_scrub(shift) if (@_);
    return $self->{value};
}

sub expires {
    my $self = shift;
    $self->{expires} = shift if (@_);
    return $self->{expires};
}

sub max_age {
    my $self = shift;
    $self->{max_age} = _cookie_scrub(shift) if (@_);
    return $self->{max_age};
}

sub path {
    my $self = shift;
    $self->{path} = _cookie_scrub(shift) if (@_);
    return $self->{path};
}

sub domain {
    my $self = shift;
    $self->{domain} = _cookie_scrub(shift) if (@_);
    return $self->{domain};
}


sub expire_cookie() {
    my $self = shift;
    $self->{expires} = 1;
}

sub _as_string {
    my $self = shift;

    unless(defined($self->{name}) and defined($self->{value})) {
        croak('Cookie Name and Value must be defined');
    }

    my $str = $self->{name} .'='. $self->{value}.';';
    $str   .= ' Path=' . $self->{path} .';'      if ($self->{path});
    $str   .= ' Domain=' . $self->{domain} . ';' if ($self->{domain});
    $str   .= ' Max-Age=' . $self->{max_age} . ';'
        if (defined($self->{max_age}));
    $str   .= ' Expires=' . _cookie_date($self->{expires}) . ';' 
        if (defined($self->{expires}));
    $str   .= ' Version=1';
    return $str;
}


# ----------------------------------------------------------------------
# Private/Utility function.
# ----------------------------------------------------------------------

#
# '=' and ';' are not valid in the name or data components of cookies
#
sub _cookie_scrub {
  my $str = shift;
  $str =~ s/(\;|\=)//g if(defined($str));
  return $str;
}


#
# this routine accepts the number of seconds to add to the server
# time to calculate the expiration string for the cookie. Cookie
# time is ALWAYS GMT!
#
sub _cookie_date() {
  my $when = shift;

  my %mn = ('Jan','01', 'Feb','02', 'Mar','03', 'Apr','04',
            'May','05', 'Jun','06', 'Jul','07', 'Aug','08',
            'Sep','09', 'Oct','10', 'Nov','11', 'Dec','12' );

  my $sydate = gmtime($when);
  my ($day, $month, $num, $time, $year) = split(/\s+/,$sydate);

  if (length($num) == 1) { 
    $num = "0$num";
  }

  my $retdate="$day $num-$month-$year $time GMT";

  return $retdate;
}


1;
__END__

=head1 NAME

NCGI::Cookie - HTTP Cookie object for NCGI

=head1 SYNOPSIS

  use NCGI::Cookie;

  # subroutine for reading cookies from the query
  my $hashref      = NCGI::Cookie::fetch();
  my $cookie_value = $hashref->{cookie_name};

  # object methods for creating cookies to send in the response
  my ($name, $value, $epoch, $path, $domain) =
    ('c1','value',1221397599,'/',undef);
  my $c = NCGI::Cookie->new(
    name    => $name,
    value   => $value,
    expires => $epoch,
    path    => $path,
    domain  => $domain,
  );

  # You can also "print $c->_as_string;"
  print $c; # c1=value; expires=Tue 07-Jun-2005 12:02:01 GMT; path=/


=head1 DESCRIPTION

B<NCGI::Cookie> provides a simple HTTP Cookie object for Cookie creation,
and a fetch subroutine to retrieve cookies sent by a browser.

=head1 SUBROUTINES

=head2 fetch

Returns a HASH reference containing all Cookies sent by the browser.

=head1 METHODS

=head2 new( ... )

Create a new NCGI::Cookie object with parameters 'name', 'value',
'expires', 'path' and 'domain'. Only 'name' and 'value' are really
necessary as expires will automatically be filled in otherwise.

=head2 name

Set or Get the name of the cookie

=head2 value

Set or Get the value of the cookie. Will be scrubbed for illegal characters.

=head2 expires

Set or Get the expiration value of the cookie as a unix epoch.

=head2 max_age

Set or Get the expiration value of the maximum age of the cookie
in seconds.

=head2 path

Set or Get the path of the cookie

=head2 domain

Set or Get the domain of the cookie

=head2 expire_cookie

Sets the expiration date of the cookie to 24 hours ago, effectively
removing it from the browser cache.

=head2 _as_string

Returns the string representation of the cookie. The "" operator is
overloaded so if you just 'print $cookie' you should also get the string
representation.

=head1 SEE ALSO

L<NCGI::Query>

=head1 AUTHOR

Mark Lawrence E<lt>nomad@null.netE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2005 Mark Lawrence E<lt>nomad@null.netE<gt>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

=cut
